File: sdk-ts/src/trading.ts
import type {
  MatchState,
  Position,
  PositionSide,
  TradeAction,
  TradeRequest,
  TradeSide,
} from './types';

/**
 * Trading strategy helper
 */
export class TradingStrategy {
  agentId: number;
  maxPositionSize: number;
  maxLeverage: number;
  riskPerTrade: number;

  constructor(agentId: number) {
    this.agentId = agentId;
    this.maxPositionSize = 5000;
    this.maxLeverage = 3;
    this.riskPerTrade = 0.1; // 10% of balance per trade
  }

  /**
   * Get current agent state from match state
   */
  private getAgentState(state: MatchState) {
    return state.agent1State.agentId === this.agentId
      ? state.agent1State
      : state.agent2State;
  }

  /**
   * Get opponent state from match state
   */
  private getOpponentState(state: MatchState) {
    return state.agent1State.agentId === this.agentId
      ? state.agent2State
      : state.agent1State;
  }

  /**
   * Calculate position size based on risk
   */
  calculateSize(balance: number, leverage: number): number {
    const riskAmount = balance * this.riskPerTrade;
    const size = riskAmount * leverage;
    return Math.min(size, this.maxPositionSize);
  }

  /**
   * Check if we have an open position in a symbol
   */
  hasPosition(state: MatchState, symbol: string): boolean {
    return this.getAgentState(state).positions.some((p) => p.symbol === symbol);
  }

  /**
   * Get current position for a symbol
   */
  getPosition(state: MatchState, symbol: string): Position | undefined {
    return this.getAgentState(state).positions.find((p) => p.symbol === symbol);
  }

  /**
   * Get current balance
   */
  getBalance(state: MatchState): number {
    return this.getAgentState(state).balance;
  }

  /**
   * Get current P&L
   */
  getPnl(state: MatchState): number {
    return this.getAgentState(state).pnl;
  }

  /**
   * Get opponent's P&L
   */
  getOpponentPnl(state: MatchState): number {
    return this.getOpponentState(state).pnl;
  }

  /**
   * Create a trade request to open a long position
   */
  openLong(symbol: string, size: number, leverage: number): TradeRequest {
    return {
      agentId: this.agentId,
      symbol,
      action: 'Open' as TradeAction,
      side: 'Long' as TradeSide,
      sizeUsd: Math.min(size, this.maxPositionSize),
      leverage: Math.min(leverage, this.maxLeverage),
    };
  }

  /**
   * Create a trade request to open a short position
   */
  openShort(symbol: string, size: number, leverage: number): TradeRequest {
    return {
      agentId: this.agentId,
      symbol,
      action: 'Open' as TradeAction,
      side: 'Short' as TradeSide,
      sizeUsd: Math.min(size, this.maxPositionSize),
      leverage: Math.min(leverage, this.maxLeverage),
    };
  }

  /**
   * Create a trade request to close a position
   */
  closePosition(symbol: string): TradeRequest {
    return {
      agentId: this.agentId,
      symbol,
      action: 'Close' as TradeAction,
      side: 'Long' as TradeSide, // Doesn't matter for close
      sizeUsd: 0,
    };
  }

  /**
   * Calculate total unrealized P&L
   */
  totalUnrealizedPnl(state: MatchState): number {
    return this.getAgentState(state).positions.reduce(
      (sum, p) => sum + p.unrealizedPnl,
      0
    );
  }

  /**
   * Check if we're winning
   */
  isWinning(state: MatchState): boolean {
    return this.getPnl(state) > this.getOpponentPnl(state);
  }

  /**
   * Get all open positions
   */
  getPositions(state: MatchState): Position[] {
    return this.getAgentState(state).positions;
  }
}

/**
 * Simple momentum signal calculator
 */
export function calculateMomentum(
  prices: Array<{ price: number; timestamp: number }>,
  lookback: number
): number {
  if (prices.length < lookback + 1) {
    return 0;
  }

  const current = prices[prices.length - 1].price;
  const past = prices[prices.length - lookback - 1].price;

  if (past === 0) {
    return 0;
  }

  return ((current - past) / past) * 100; // Return as percentage
}

/**
 * Simple moving average calculator
 */
export function calculateSMA(prices: number[], period: number): number | null {
  if (prices.length < period) {
    return null;
  }

  const slice = prices.slice(-period);
  const sum = slice.reduce((a, b) => a + b, 0);
  return sum / period;
}

/**
 * Exponential moving average calculator
 */
export function calculateEMA(prices: number[], period: number): number | null {
  if (prices.length < period) {
    return null;
  }

  const multiplier = 2 / (period + 1);
  let ema = prices.slice(0, period).reduce((a, b) => a + b, 0) / period;

  for (let i = period; i < prices.length; i++) {
    ema = (prices[i] - ema) * multiplier + ema;
  }

  return ema;
}

/**
 * Relative Strength Index calculator
 */
export function calculateRSI(prices: number[], period: number = 14): number | null {
  if (prices.length < period + 1) {
    return null;
  }

  const changes: number[] = [];
  for (let i = 1; i < prices.length; i++) {
    changes.push(prices[i] - prices[i - 1]);
  }

  const recentChanges = changes.slice(-period);
  const gains = recentChanges.filter((c) => c > 0);
  const losses = recentChanges.filter((c) => c < 0).map((c) => Math.abs(c));

  const avgGain = gains.length > 0 ? gains.reduce((a, b) => a + b, 0) / period : 0;
  const avgLoss = losses.length > 0 ? losses.reduce((a, b) => a + b, 0) / period : 0;

  if (avgLoss === 0) {
    return 100;
  }

  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

File: sdk/src/trading.rs
use crate::models::{MatchState, Position, PositionSide, TradeAction, TradeRequest, TradeSide};

/// Trading strategy helper
pub struct TradingStrategy {
    pub agent_id: u64,
    pub max_position_size: f64,
    pub max_leverage: f64,
    pub risk_per_trade: f64, // As fraction of balance
}

impl TradingStrategy {
    pub fn new(agent_id: u64) -> Self {
        Self {
            agent_id,
            max_position_size: 5000.0,
            max_leverage: 3.0,
            risk_per_trade: 0.1, // 10% of balance per trade
        }
    }

    /// Calculate position size based on risk
    pub fn calculate_size(&self, balance: f64, leverage: f64) -> f64 {
        let risk_amount = balance * self.risk_per_trade;
        let size = risk_amount * leverage;
        size.min(self.max_position_size)
    }

    /// Check if we have an open position in a symbol
    pub fn has_position(&self, state: &MatchState, symbol: &str) -> bool {
        let agent_state = if state.agent1_state.agent_id == self.agent_id {
            &state.agent1_state
        } else {
            &state.agent2_state
        };

        agent_state.positions.iter().any(|p| p.symbol == symbol)
    }

    /// Get current position for a symbol
    pub fn get_position<'a>(&self, state: &'a MatchState, symbol: &str) -> Option<&'a Position> {
        let agent_state = if state.agent1_state.agent_id == self.agent_id {
            &state.agent1_state
        } else {
            &state.agent2_state
        };

        agent_state.positions.iter().find(|p| p.symbol == symbol)
    }

    /// Get current balance
    pub fn get_balance(&self, state: &MatchState) -> f64 {
        if state.agent1_state.agent_id == self.agent_id {
            state.agent1_state.balance
        } else {
            state.agent2_state.balance
        }
    }

    /// Get current P&L
    pub fn get_pnl(&self, state: &MatchState) -> f64 {
        if state.agent1_state.agent_id == self.agent_id {
            state.agent1_state.pnl
        } else {
            state.agent2_state.pnl
        }
    }

    /// Create a trade request to open a long position
    pub fn open_long(&self, symbol: &str, size: f64, leverage: f64) -> TradeRequest {
        TradeRequest {
            agent_id: self.agent_id,
            symbol: symbol.to_string(),
            action: TradeAction::Open,
            side: TradeSide::Long,
            size_usd: size.min(self.max_position_size),
            leverage: Some(leverage.min(self.max_leverage)),
        }
    }

    /// Create a trade request to open a short position
    pub fn open_short(&self, symbol: &str, size: f64, leverage: f64) -> TradeRequest {
        TradeRequest {
            agent_id: self.agent_id,
            symbol: symbol.to_string(),
            action: TradeAction::Open,
            side: TradeSide::Short,
            size_usd: size.min(self.max_position_size),
            leverage: Some(leverage.min(self.max_leverage)),
        }
    }

    /// Create a trade request to close a position
    pub fn close_position(&self, symbol: &str) -> TradeRequest {
        TradeRequest {
            agent_id: self.agent_id,
            symbol: symbol.to_string(),
            action: TradeAction::Close,
            side: TradeSide::Long, // Side doesn't matter for close
            size_usd: 0.0,         // Will close entire position
            leverage: None,
        }
    }

    /// Calculate total unrealized P&L
    pub fn total_unrealized_pnl(&self, state: &MatchState) -> f64 {
        let agent_state = if state.agent1_state.agent_id == self.agent_id {
            &state.agent1_state
        } else {
            &state.agent2_state
        };

        agent_state
            .positions
            .iter()
            .map(|p| p.unrealized_pnl)
            .sum()
    }

    /// Get opponent's P&L for comparison
    pub fn opponent_pnl(&self, state: &MatchState) -> f64 {
        if state.agent1_state.agent_id == self.agent_id {
            state.agent2_state.pnl
        } else {
            state.agent1_state.pnl
        }
    }
}

/// Simple momentum signal calculator
pub fn calculate_momentum(prices: &[(f64, u64)], lookback: usize) -> f64 {
    if prices.len() < lookback + 1 {
        return 0.0;
    }

    let current = prices.last().map(|(p, _)| *p).unwrap_or(0.0);
    let past = prices
        .get(prices.len() - lookback - 1)
        .map(|(p, _)| *p)
        .unwrap_or(current);

    if past == 0.0 {
        return 0.0;
    }

    (current - past) / past * 100.0 // Return as percentage
}

/// Simple moving average calculator
pub fn calculate_sma(prices: &[f64], period: usize) -> Option<f64> {
    if prices.len() < period {
        return None;
    }

    let sum: f64 = prices.iter().rev().take(period).sum();
    Some(sum / period as f64)
}
